### DISCUSSION TITLE
Best place to initialize a DB connection

### DISCUSSION THREAD

**Discussion by ChristophP** [vikejs/vike#1940] [Discussion]
Hi,

# Short story
We want to have HMR in our monorepo, therefore we want to include many deps in the SSR bundle (to get HMR https://vite.dev/config/ssr-options.html#ssr-external). That means moving DB connecting logic from the Express app into the Vike code.

What the best place to put things like connecting to a DB?

### Places we've tried
1. In a file that is built by Vike/Vite (`renderer/+onBeforeRender`) on the top-level (so that it connect upon import) with a top level await -> gives an esbuild error saying that to top level await is not legal for the targets `chrome87, safari 14, ....` (apparently it is trying to build for client which is surprising since `+onBeforeRender` is a server only hook)
2. Inside `onBeforeRender` -> not a great place since we need to connect before the requests come
3. There's an issue for `onBoot` but it doesn't seem to be implemented yet

What's the general recommendation for this sort of thing?
We appreciate some pointers.

**Comment by brillout** [comment]
Hi Dido,

have you tried something like that?

```ts
async function getConnection() {
  if (!globalThis.connection) globalThis.connection = await initConnection()
  return globalThis.connection
}
```

> it is trying to build for client which is surprising

That's determined by Vite.

**Reply by ChristophP** [reply]
Hi sorry for the late reply.
Thanks for the hint with globalThis
